### ISSUE TITLE
Unet not respecting text conditioning

### ISSUE THREAD

**Issue by marunine** [lucidrains/imagen-pytorch#53]
I've done a few test runs with your Imagen and DALLE2 models on some smaller scale models (~192 dim) on a 10M dataset across 30 epochs.

I've found that the DALLE2 model guided by CLIP image and text embeds has better alignment than the Imagen model, and that the Imagen model never seems to really get any sort of alignment, almost like it's stuck doing unconditional generation.

Looking through the code I wonder if it's something to do with how text embeds are handled in general and that the DALLE2 model just gets away with using the Image embeddings.

I wonder if there might be a problem with this:

https://github.com/lucidrains/imagen-pytorch/blob/ba50fc8c5dd8e651b87cab578cf7752bcc34ec91/imagen_pytorch/imagen_pytorch.py#L1119

The null text embed is filled with random values per your initialization:

https://github.com/lucidrains/imagen-pytorch/blob/ba50fc8c5dd8e651b87cab578cf7752bcc34ec91/imagen_pytorch/imagen_pytorch.py#L937

...and these values persist with the rest of the model after construction time.

The model seems to be allowed to look at the entire space of the embedding, however:

https://github.com/lucidrains/imagen-pytorch/blob/ba50fc8c5dd8e651b87cab578cf7752bcc34ec91/imagen_pytorch/imagen_pytorch.py#L1132

There's no way for the model to understand that the random data is different from the rest of the embedding as far as I can tell since the masks are only used to decide whether the random data is used instead of the actual token embedding data.

I'm going to try doing a run without using the token_mask and just padding to max length to try and avoid picking up the null_text_embed, but for pad to longest I'm not sure this is working as intended.

**Comment by lucidrains** [comment]
@marunine hello! this is actually exactly how dalle2 (and dalle1) handled this (instead of masking out, they attend to learned padding tokens, and they act like a sort of like memory key / values) do you want to see if https://github.com/lucidrains/imagen-pytorch/commit/9db3cba4d61c3e593e82b435c845dc405a49c389 helps?

**Comment by marunine** [comment]
I must have missed that part in the paper, thanks for pointing that out!

Though at the same time, I thought they found that text conditioning didn't help in their experiments - the opposite of what Google found with Imagen. Didn't DALLE2 just use image embeddings for both the base and upscaler Unets? It makes me wonder if there was some problem with the implementation or approach. Intuitively I think it's making the conditioning a lot harder for the Unet if the Unet can't tell what's padded or not and gets real looking data for all its samples, but randomly based on padding length.

I'll give your update a shot anyway and report back on the results.

**Comment by lucidrains** [comment]
@marunine oh oops, sorry, i meant just DALLE1 my bad

but if you are seeing good results on DALLE2 using my repository, then it is also using the same attention to padding tokens. that's what i meant to say lol

**Comment by s4sarath** [comment]
What instead of null k v, let's say we set for eg '_' as fixed token for text encoder, for unconditional cases. So, that eventually model sees this particular token with all images, it might learn that this token is s generic one, which is equivalent to unconditional.  Does it make sense. I am currently implementing it in TF2. Not complete.

**Comment by lucidrains** [comment]
@marunine @s4sarath what i can do is just offer the version where the cross attention masks out properly. i also plan on building linear attention for the cross attention layers that are not computationally feasible, and that should strengthen the conditioning even more

**Comment by lucidrains** [comment]
> I must have missed that part in the paper, thanks for pointing that out!
> 
> Though at the same time, I thought they found that text conditioning didn't help in their experiments - the opposite of what Google found with Imagen. Didn't DALLE2 just use image embeddings for both the base and upscaler Unets? It makes me wonder if there was some problem with the implementation or approach. Intuitively I think it's making the conditioning a lot harder for the Unet if the Unet can't tell what's padded or not and gets real looking data for all its samples, but randomly based on padding length.
> 
> I'll give your update a shot anyway and report back on the results.

to address your second paragraph, you are correct. google claims that scaling the text encoder is enough for text conditioning

**Comment by lucidrains** [comment]
@Netruk44 have you observed the same results as @marunine ?

**Comment by Netruk44** [comment]
I noticed the same behavior when calling `sample` with a `cond_scale` around `1.0`. When I increased the value to around 3, the images began to align better with the prompts, and a value of around 5-10 I noticed 'optimal' alignment. Values above 10 started to fall into a kind of overfitting where the resulting images mostly looked the same.

Changing `cond_drop_prob` to `0.1` partway through training did seem to help with alignment a bit, but a `cond_scale` of `1.0` still wasn't aligned very well.

**Comment by lucidrains** [comment]
> I noticed the same behavior when calling `sample` with a `cond_scale` around `1.0`. When I increased the value to around 3, the images began to align better with the prompts, and a value of around 5-10 I noticed 'optimal' alignment. Values above 10 started to fall into a kind of overfitting where the resulting images mostly looked the same.
> 
> Changing `cond_drop_prob` to `0.1` partway through training did seem to help with alignment a bit, but a `cond_scale` of `1.0` still wasn't aligned very well.

great to hear! so it is working!

**Comment by Netruk44** [comment]
> I thought that the model throws an assertion when `cond_scale` isn't set to `1.0` if it isn't unconditional?

Sorry, I meant `cond_scale` when calling `sample`: https://github.com/lucidrains/imagen-pytorch/blob/ad4c576988754dd74ecf3291a5fe2eb5375c8cad/imagen_pytorch/imagen_pytorch.py#L1457

**Comment by lucidrains** [comment]
@marunine no worries, i'll write up a small section on the `cond_scale` and explain what classifier free guidance is doing

**Comment by lucidrains** [comment]
i'm going to further strengthen the conditioning now with linear cross attention layers even if full cross attention is prohibitive and turned off

**Comment by Netruk44** [comment]
For examples of alignment in my model, here's some outputs from the first unet in the cascade (The other parts of the cascade still aren't working so well for me yet).

(Just for caveats, it's intended as a very limited toy model, so prompts such as `armchair in the shape of an avocado` are outside the realm of my model's capabilities)

`jungle`:
![jungle image](https://cdn.discordapp.com/attachments/388772569823576065/984527006236540939/unknown.png)

`indoors, inside`:
![indoors image](https://cdn.discordapp.com/attachments/388772569823576065/984527005838094366/unknown.png)

`desert`:
![desert image](https://cdn.discordapp.com/attachments/979781255270920233/984522638212857956/unknown.png)

**Comment by marunine** [comment]
I think the assertion I saw was when trying the `cond_scale` change in sample in the DALLE2 repo.

I'm seeing good alignment on my model with a scale of `3.0`. Thanks for the tips!

**Comment by lucidrains** [comment]
https://github.com/lucidrains/imagen-pytorch#faq @Netruk44 @marunine done

**Comment by lucidrains** [comment]
@marunine no, thank you! this was very helpful

as you can tell, i'm quick at building stuff, but lack the time to sit down and train (which is equally important). so you're reporting of results really helps me :smile:
